### PR TITLE
Add compression [Part 1]

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
In regards to #492

There will be more parts to compression, this is just the first.

### CODE REVIEW

File: core/src/main/java/io/grpc/MessageEncoding.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  /**
+   * Represents a message compressor.
+   */
+  public interface Compressor {
+    /**
+     * Returns the message encoding that this compressor uses.
+     *
+     * <p>This can be values such as "gzip", "deflate", "snappy", etc.
+     */
+    String getMessageEncoding();
```
Comment (nmittler) [inline core/src/main/java/io/grpc/MessageEncoding.java]: Maybe a super interface `MessageEncoder`?

File: core/src/main/java/io/grpc/MessageEncoding.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
+package io.grpc;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * TODO: Insert description here. (generated by notcarl)
```
Comment (nmittler) [inline core/src/main/java/io/grpc/MessageEncoding.java]: better comment? :)

File: core/src/main/java/io/grpc/CallOptions.java [hunk 3]
```diff
@@ -86,6 +91,23 @@ public Long getDeadlineNanoTime() {
     return deadlineNanoTime;
   }
 
+  /**
+   * Retruns the Codec used for compression and decompression, or {@code null} if none is set.
+   */
+  @Nullable
+  public Codec getCompression() {
```
Comment (nmittler) [inline core/src/main/java/io/grpc/CallOptions.java]: `getCompressionCodec`?

File: core/src/main/java/io/grpc/CallOptions.java [hunk 4]
```diff
@@ -86,6 +91,23 @@ public Long getDeadlineNanoTime() {
...
+    return codec;
+  }
+
+  /**
+   * Use the desired compression.
+   */
+  public CallOptions withCompression(@Nullable Codec codec) {
```
Comment (nmittler) [inline core/src/main/java/io/grpc/CallOptions.java]: `withCodec`?

File: core/src/main/java/io/grpc/MessageEncoding.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    public OutputStream compress(OutputStream os) throws IOException {
+      return os;
+    }
+  };
+
+
+  /**
+   * Represents a message compressor.
+   */
+  public interface Compressor {
```
Comment (nmittler) [inline core/src/main/java/io/grpc/MessageEncoding.java]: Do we benefit by having `Compressor` and `Decompressor` as separate interfaces? Would it be simpler to just have `MessageEncoder`?

File: core/src/main/java/io/grpc/inprocess/InProcessTransport.java [hunk 6]
```diff
@@ -205,6 +206,11 @@ private synchronized void setListener(ClientStreamListener listener) {
       }
 
       @Override
+      public void setDefaultCodec(Codec c) {
+        // nop
```
Comment (nmittler) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: Add TODO instead?

File: core/src/main/java/io/grpc/internal/Stream.java [hunk 7]
```diff
@@ -77,4 +79,10 @@
    * result in excessive buffering within the transport.
    */
   boolean isReady();
+
+  /**
+   * Sets the default message encoder for messages on this stream.
+   * @param c the compressor / decompressor
+   */
+  void setDefaultCodec(Codec c);
```
Comment (nmittler) [inline core/src/main/java/io/grpc/internal/Stream.java]: Is this the lever that we want? I would expect the default to be channel-wide ... no?

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 5 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/MessageEncoding.java]: Thinking out loud:

One of the ideas I had for how this would work would be to add methods to different compressors.  For example, Gzip could have a setCompressionLevel method on it that would be specific to it (and perhaps not snappy, or other compression type).  Using such a method would be specific to compressing, but not decompressing, so it would look odd.  Something I want to avoid:
```
// This doesn't do what you think it does
codec.setCompressionLevel(Level.FAST).decode(message);
```

Specifically, it is possible to imagine wanting to set encode or decode specific methods that would not apply to the other.  While not the end of the world, it would add to the wartiness of the API.  

It would also make testing /slightly/ easier since framers and deframers only care about implementing one of those methods.

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 2 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/MessageEncoding.java]: Done

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 1 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/MessageEncoding.java]: I had a hard time thinking of when you would want just a MessageEncoder string.  Did you have a use case in mind?

File: core/src/main/java/io/grpc/CallOptions.java (hunk 3 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/CallOptions.java]: getCompressionCodec seems like a decent name.  Codec is a COder / DECoder, but I have been using it as COmpressor / DECompressor.  Since this would be on the public API I tried to pick a name that a user would think of ("I want to turn on compression, let me look for methods with the word compression in them").  getCompressionCodec would fit that.

My one qualm is that it would be like an ATM Machine: slightly redundant in its name.

File: core/src/main/java/io/grpc/CallOptions.java (hunk 4 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/CallOptions.java]: Updated in regards to the above comment.

File: core/src/main/java/io/grpc/CallOptions.java [hunk 8]
```diff
@@ -104,6 +127,7 @@ public String toString() {
       buffer.append(" (").append(deadlineNanoTime - System.nanoTime()).append(" ns from now)");
     }
     buffer.append("]");
+    buffer.append("\ncodec=").append(codec);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/CallOptions.java]: I was thinking the [] was supposed to enclose all the options, so that we would have `[deadlineNanoTime=12341234 (50 ns from now), codec=SOMETHING]`

File: core/src/main/java/io/grpc/inprocess/InProcessTransport.java (hunk 6 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: This was actually intentional.  For the inprocess transport, there is no Framer or deframer, so it would be a little wonky to set a codec.  If you think there might be a use for enabling coding on the in process I would be willing to add something in.

File: core/src/main/java/io/grpc/internal/Stream.java (hunk 7 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/Stream.java]: Now that you ask.... I suppose not.  The spec says that you should be able to request that the whole channel use a compressor, ask that an individual call use a compressor different than the channel default, and also to be able to selectively turn off compression for each message.  

Currently the presence of a Codec implies that compression is desired even those these two ideas are separable.  Because this feature is being added in chunks, I personally expected this API to change slightly with each commit until it is ready to be released.   (it is easier to see problems once the entire thing is in place, rather than at this early stage)

File: core/src/main/java/io/grpc/internal/MessageDeframer.java [hunk 9]
```diff
@@ -346,19 +349,15 @@ private InputStream getUncompressedBody() {
...
-    if (compression != Compression.GZIP) {
-      throw new AssertionError("Unknown compression type");
-    }
-
     try {
-      return new GZIPInputStream(ReadableBuffers.openStream(nextFrame, true));
+      return decompressor.decompress(ReadableBuffers.openStream(nextFrame, true));
     } catch (IOException e) {
-      throw new RuntimeException(e);
+      throw Throwables.propagate(e);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: `RuntimeException` was much more explicit. `Throwables.propagate` will just make a `RuntimeException` itself. Why the change?

File: core/src/main/java/io/grpc/internal/MessageDeframer.java [hunk 10]
```diff
@@ -346,19 +349,15 @@ private InputStream getUncompressedBody() {
...
           "Can't decode compressed frame as compression not configured.").asRuntimeException();
     }
 
-    if (compression != Compression.GZIP) {
-      throw new AssertionError("Unknown compression type");
-    }
-
     try {
-      return new GZIPInputStream(ReadableBuffers.openStream(nextFrame, true));
+      return decompressor.decompress(ReadableBuffers.openStream(nextFrame, true));
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: This is dangerous, as we are running application code on the network thread.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java [hunk 11]
```diff
@@ -346,19 +349,15 @@ private InputStream getUncompressedBody() {
   }
 
   private InputStream getCompressedBody() {
-    if (compression == Compression.NONE) {
+    if (decompressor == MessageEncoding.NONE) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: This just seems... off. Surely this isn't appropriate?

File: core/src/main/java/io/grpc/CallOptions.java [hunk 12]
```diff
@@ -86,6 +91,23 @@ public Long getDeadlineNanoTime() {
     return deadlineNanoTime;
   }
 
+  /**
+   * Retruns the Codec used for compression and decompression, or {@code null} if none is set.
```
Comment (ejona86) [inline core/src/main/java/io/grpc/CallOptions.java]: This makes it sound like the response must be using the same compression type if using compression. However, the server is permitted to choose a different compression algorithm if the client supports it. How is `Message-Accept-Encoding` going to play into this?

File: core/src/main/java/io/grpc/MessageEncoding.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+     * @param is The input stream of uncompressed data
+     * @return An input stream that decompresses
+     */
+    InputStream decompress(InputStream is) throws IOException;
+  }
+
+  /**
+   * Represents an object that can both compress and decompress messages.
+   */
+  public interface Codec extends Compressor, Decompressor {}
```
Comment (ejona86) [inline core/src/main/java/io/grpc/MessageEncoding.java]: Where is `Compressor` and `Decompressor` going? Only `Codec` is usable right now. When will the interfaces be used separately?

File: core/src/main/java/io/grpc/MessageEncoding.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+  /**
+   * A gzip compressor and decompressor.  In the future this will likely support other
+   * decompression methods, such as compression level.
```
Comment (ejona86) [inline core/src/main/java/io/grpc/MessageEncoding.java]: How is "compression level" a "decompression method"?

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 5 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/MessageEncoding.java]: But if a `Codec` is required for `CallOptions`, then using the interfaces separately doesn't seem possible.

File: core/src/main/java/io/grpc/CallOptions.java (hunk 3 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/CallOptions.java]: We _need_ compression in the name, otherwise it seems like a marshaller. I would fully expect a `MessageEncoding.Codec` to be "proto" or similar.

File: core/src/main/java/io/grpc/internal/Stream.java (hunk 7 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Stream.java]: What does it mean to be the "default codec"? What is "default" about it?

File: core/src/main/java/io/grpc/CallOptions.java (hunk 8 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/CallOptions.java]: Perhaps the ToStringBuilder should be used instead?

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 9 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: Increase consistency across the code base.  Other places in the code already do this as precedent.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 10 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: This is the same as it was before.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 11 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: Not sure what you mean.

File: core/src/main/java/io/grpc/CallOptions.java (hunk 12 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/CallOptions.java]: As this is a work in progress, it will likely change.  This is just *something* that works now.

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 13 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/MessageEncoding.java]: See the response to Nathan earlier in the thread.

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 14 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/MessageEncoding.java]: The decompression is a artifact of an earlier version when Gzip was only a decompressor.

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 5 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/MessageEncoding.java]: Codec isn't required, additional withMethods can be added.  It may or may not be a good idea to allow this, but it is certainly possible.

File: core/src/main/java/io/grpc/internal/Stream.java (hunk 7 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/Stream.java]: The codec to use by default if a message is compressed (and not used if not compressed).

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 9 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: I'm actually trying to remove those, instead of adding more. I've had very good success with looking for Throwables.propagate() to find exception handling bugs; the tool itself isn't the problem, but its overuse has led to ignoring errors.

Still, is it really "consistency"? Can you find a place we catch a `Throwable` other than `Throwable` or `Exception` and use `Throwables.propagate()`? Those are the only cases where `Throwables.propagate()` seems useful, in order to prevent needless exception-wrapping.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 10 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: No...?? Before we were always using GZIPInputStream. Now we are calling a method provided by the application. What am I missing?

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 11 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: I guess I'm having trouble with the definition of NONE:
> Special sentinel codec indicating that no compression should be used.  Users should use reference equality to see if compression is disabled.

I guess it's fine. I think I see the reasons, but something just seems... off.

File: core/src/main/java/io/grpc/MessageEncoding.java [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+   */
+  public static final Codec NONE = new Codec() {
+    @Override
+    public InputStream decompress(InputStream is) throws IOException {
+      return is;
+    }
+
+    @Override
+    public String getMessageEncoding() {
+      return null;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/MessageEncoding.java]: Return `"identity"`?

File: core/src/main/java/io/grpc/internal/Stream.java (hunk 7 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Stream.java]: But what is "default" about it. Why is it not `setCodec`? With "default," I expect it to be overridden with another `Codec` at some other point.

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 15 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/MessageEncoding.java]: +1

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 5 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/MessageEncoding.java]: @carl-mastrangelo 

> Using such a method would be specific to compressing, but not decompressing, ...

If that's the case, then I would expect `Codec` to use composition:

```java
interface Codec {
    String name();
    Compressor compressor();
    Decompressor decompressor();
}

interface Compressor {
    OutputStream compress(OutputStream os) throws IOException;
}

interface Decompressor {
    InputStream decompress(InputStream is) throws IOException;
}
```

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 1 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/MessageEncoding.java]: I think the `Codec` is the only thing that needs the name ... probably just leave it out of the compressor/decompressor.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 9 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: Yeah, IIRC there is an effort across g3 to remove references to `Throwables.propagate` as well.

File: core/src/main/java/io/grpc/internal/Stream.java (hunk 7 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/internal/Stream.java]: I think this is the main place where we need a bit of design.  How do we configure the system to use codecs ... and what use cases do we support, etc.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 10 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: The expected case is that decompress takes an input stream, and returns another input stream, not that it does any actual processing itself.  I can give it a more clear name if you would like.

File: core/src/main/java/io/grpc/internal/MessageDeframer.java (hunk 11 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/MessageDeframer.java]: Using a sentinel object has some benefits over using null.  The best one IMO is being easy to find references to it.  It also buys a little extra meaning in the context of nullable things, notably an explicitly "empty" thing, vs an implicitly empty thing, such as null.  It also provides a a hook to providing clarity in the error message ("You should not have called me")  vs getting an NPE.  This code does such as check, but the call site has to make it rather than putting that logic in the object itself.


On a tangential note, I could feasibly see the error below being caused by the remote side (not setting any compression, and then sending a compressed message).

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 15 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/MessageEncoding.java]: @ejona86 & @nmittler null is special here, its the type that is expected to be set in the Message-Encoding and Message-Accept-Encoding headers.  I don't see the word "identity" in the grpc spec.

Comment (ejona86) [inline core/src/main/java/io/grpc/MessageEncoding.java]: @dgquintas had added "none" to the internal specification doc, not realizing that spec had been superseded. I then created grpc/grpc-common#257 to get "identity" added after discussing with him. Like many fixes to the spec, it has languished. There was one reference to "none" in the compression doc that should have been updated to "identity" but was missed (David just fixed it).

It behaves like Content-Encoding "identity"; "identity" is always implicitly supported and there is no need to send it on the wire.

File: core/src/main/java/io/grpc/CallOptions.java (hunk 8 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/CallOptions.java]: +1, will that add a guava dependency or do we already have it from this module?

Comment (ejona86) [inline core/src/main/java/io/grpc/CallOptions.java]: We're already using `ToStringHelper` in `AbstractClientStream`.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/CallOptions.java]: We also use Preconditions extensively.  If we care about breaking the dep, we can clone the library and cherry pick in functions.

Comment (nmittler) [inline core/src/main/java/io/grpc/CallOptions.java]: Ok, so ToStringBuilder it is :)

**Comment by carl-mastrangelo** [comment]
@ejona86 & @nmittler  PTAL

File: core/src/main/java/io/grpc/ChannelImpl.java [hunk 16]
```diff
@@ -376,6 +397,10 @@ public void transportTerminated() {
   public static final Metadata.Key<Long> TIMEOUT_KEY =
       Metadata.Key.of(HttpUtil.TIMEOUT, new TimeoutMarshaller());
 
+  // TODO(carl-mastrangelo): move this to internal
+  public static final Metadata.Key<String> MESSAGE_ENCODING_KEY =
+      Metadata.Key.of(HttpUtil.MESSAGE_ENCODING, Metadata.ASCII_STRING_MARSHALLER);
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ChannelImpl.java]: Should we add this to HttpUtil (under internal)?  We should probably rename it to `GrpcUtil` since it's basically set of all purpose constants.  I've actually done this in #680.

Comment (nmittler) [inline core/src/main/java/io/grpc/ChannelImpl.java]: Ah I see you already had a TODO for this ... so I gave you the appropriate place to put it :)

**Comment by nmittler** [comment]
@carl-mastrangelo LGTM as a first pass ... good work!  We can always iterate if we decide to rearrange things.

File: core/src/main/java/io/grpc/internal/AbstractClientStream.java [hunk 17]
```diff
@@ -116,6 +117,10 @@ protected void inboundHeadersReceived(Metadata.Headers headers) {
       log.log(Level.INFO, "Received headers on closed stream {0} {1}",
           new Object[]{id(), headers});
     }
+    if (headers.containsKey(ChannelImpl.MESSAGE_ENCODING_KEY)) {
+      String messageEncoding = headers.get(ChannelImpl.MESSAGE_ENCODING_KEY);
+      setDecompressor(messageEncoding);
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/AbstractClientStream.java]: @ejona86 : question about this line:  is it safe to throw an exception from here?   If the stream has access to the listener, should the stream be closed, the listener contacted, and this method complete normally?

File: core/src/main/java/io/grpc/MessageEncoding.java (hunk 5 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/MessageEncoding.java]: @nmittler Compressor and Decompressor are pretty tied to their message encoding name.  Both need their name in order to be self describing.

Codec is present out of convenience, not necessity.  The composition over inheritance argument doesn't apply here, since that is for implementation inheritance, not interface inheritance.  With the latest changes, Codec is currently only used in this class and not used elsewhere.  It is exposed here for client convenience.

Comment (nmittler) [inline core/src/main/java/io/grpc/MessageEncoding.java]: @carl-mastrangelo Don't have a strong preference one way or another. I'm fine with it as you have it.

File: core/src/main/java/io/grpc/internal/AbstractClientStream.java (hunk 17 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/internal/AbstractClientStream.java]: Since @nmittler  LGTM'd this, I am going to go ahead an submit.  If there are lingering issues comment on this PR and I can address them in a followup PR.

**Comment by ejona86** [comment]
This should not have been committed to master. I thought we agreed it was going to go into its own branch?

**Comment by carl-mastrangelo** [comment]
The PR changed significantly since when we talked about that.  I don't believe the reason to have a separate branch was still justified.
